File: packages/go/currency_l1_client.go
package constellation

import "fmt"

// CurrencyL1Client is a client for interacting with Currency L1 nodes
//
// Example:
//
//	config := NetworkConfig{L1URL: "http://localhost:9010"}
//	client, err := NewCurrencyL1Client(config)
//	if err != nil {
//	    return err
//	}
//
//	// Get last reference for an address
//	lastRef, err := client.GetLastReference("DAG...")
//
//	// Submit a transaction
//	result, err := client.PostTransaction(signedTx)
//
//	// Check transaction status
//	pending, err := client.GetPendingTransaction(result.Hash)
type CurrencyL1Client struct {
	client *HTTPClient
}

// NewCurrencyL1Client creates a new CurrencyL1Client
//
// Returns an error if L1URL is not provided in the config
func NewCurrencyL1Client(config NetworkConfig) (*CurrencyL1Client, error) {
	if config.L1URL == "" {
		return nil, ErrL1URLRequired
	}

	client := NewHTTPClient(config.L1URL, config.Timeout)
	return &CurrencyL1Client{client: client}, nil
}

// GetLastReference gets the last accepted transaction reference for an address
//
// This is needed to create a new transaction that chains from
// the address's most recent transaction.
func (c *CurrencyL1Client) GetLastReference(address string) (*TransactionReference, error) {
	var result TransactionReference
	path := fmt.Sprintf("/transactions/last-reference/%s", address)
	if err := c.client.Get(path, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// PostTransaction submits a signed currency transaction to the L1 network
func (c *CurrencyL1Client) PostTransaction(transaction *CurrencyTransaction) (*PostTransactionResponse, error) {
	var result PostTransactionResponse
	if err := c.client.Post("/transactions", transaction, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// GetPendingTransaction gets a pending transaction by hash
//
// Use this to poll for transaction status after submission.
// Returns nil if the transaction is not found (already confirmed or invalid).
func (c *CurrencyL1Client) GetPendingTransaction(hash string) (*PendingTransaction, error) {
	var result PendingTransaction
	path := fmt.Sprintf("/transactions/%s", hash)
	if err := c.client.Get(path, &result); err != nil {
		if netErr, ok := err.(*NetworkError); ok && netErr.StatusCode == 404 {
			return nil, nil
		}
		return nil, err
	}
	return &result, nil
}

// CheckHealth checks the health/availability of the L1 node
func (c *CurrencyL1Client) CheckHealth() bool {
	var result interface{}
	return c.client.Get("/cluster/info", &result) == nil
}

File: packages/rust/src/network/currency_l1_client.rs
//! Currency L1 client for submitting and querying transactions

use super::client::HttpClient;
use super::types::{
    NetworkConfig, NetworkError, NetworkResult, PendingTransaction, PostTransactionResponse,
};
use crate::currency_types::{CurrencyTransaction, TransactionReference};

/// Client for interacting with Currency L1 nodes
///
/// # Example
///
/// ```ignore
/// use constellation_sdk::network::{CurrencyL1Client, NetworkConfig};
///
/// let config = NetworkConfig {
///     l1_url: Some("http://localhost:9010".to_string()),
///     ..Default::default()
/// };
///
/// let client = CurrencyL1Client::new(config)?;
///
/// // Get last reference for an address
/// let last_ref = client.get_last_reference("DAG...").await?;
///
/// // Submit a transaction
/// let result = client.post_transaction(&signed_tx).await?;
///
/// // Check transaction status
/// if let Some(pending) = client.get_pending_transaction(&result.hash).await? {
///     println!("Status: {:?}", pending.status);
/// }
/// ```
pub struct CurrencyL1Client {
    client: HttpClient,
}

impl CurrencyL1Client {
    /// Create a new CurrencyL1Client
    ///
    /// # Errors
    ///
    /// Returns an error if l1_url is not provided in the config
    pub fn new(config: NetworkConfig) -> NetworkResult<Self> {
        let l1_url = config.l1_url.ok_or_else(|| {
            NetworkError::ConfigError("l1_url is required for CurrencyL1Client".into())
        })?;

        let client = HttpClient::new(l1_url, config.timeout)?;
        Ok(Self { client })
    }

    /// Get the last accepted transaction reference for an address
    ///
    /// This is needed to create a new transaction that chains from
    /// the address's most recent transaction.
    pub async fn get_last_reference(&self, address: &str) -> NetworkResult<TransactionReference> {
        self.client
            .get(&format!("/transactions/last-reference/{}", address))
            .await
    }

    /// Submit a signed currency transaction to the L1 network
    pub async fn post_transaction(
        &self,
        transaction: &CurrencyTransaction,
    ) -> NetworkResult<PostTransactionResponse> {
        self.client.post("/transactions", transaction).await
    }

    /// Get a pending transaction by hash
    ///
    /// Use this to poll for transaction status after submission.
    /// Returns None if the transaction is not found (already confirmed or invalid).
    pub async fn get_pending_transaction(
        &self,
        hash: &str,
    ) -> NetworkResult<Option<PendingTransaction>> {
        match self.client.get(&format!("/transactions/{}", hash)).await {
            Ok(tx) => Ok(Some(tx)),
            Err(NetworkError::HttpError {
                status_code: Some(404),
                ..
            }) => Ok(None),
            Err(e) => Err(e),
        }
    }

    /// Check the health/availability of the L1 node
    pub async fn check_health(&self) -> bool {
        self.client
            .get::<serde_json::Value>("/cluster/info")
            .await
            .is_ok()
    }
}
